pkg/controllers/providers/ssm/invalidation/controller.go

```diff
@@ -0,0 +1,97 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package invalidation
+
+import (
+	"context"
+	"time"
+
+	"github.com/awslabs/operatorpkg/singleton"
+	"github.com/patrickmn/go-cache"
+	"github.com/samber/lo"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+
+	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/amifamily"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/ssm"
+)
+
+const reconcilePeriod = 30 * time.Minute
+
+// The SSM Invalidation controller is responsible for invalidating "latest" SSM parameters when they point to deprecated
+// AMIs. This can occur when an EKS-optimized AMI with a regression is released, and the AMI team chooses to deprecate
+// the AMI rather than roll forward. Normally, SSM parameter cache entries expire after 24 hours to prevent a thundering
+// herd upon a new AMI release, however Karpenter should react faster when an AMI is deprecated. This controller will
+// ensure Karpenter reacts to AMI deprecations within it's polling period (30m).
+type Controller struct {
+	cache       *cache.Cache
+	amiProvider amifamily.Provider
+}
+
+func NewController(ssmCache *cache.Cache, amiProvider amifamily.Provider) *Controller {
+	return &Controller{
+		cache:       ssmCache,
+		amiProvider: amiProvider,
+	}
+}
+
+func (c *Controller) Name() string {
+	return "providers.ssm.invalidation"
+}
+
+func (c *Controller) Reconcile(ctx context.Context) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, c.Name())
+
+	entries := map[string]ssm.Parameter{}
+	for _, item := range c.cache.Items() {
```

Comment:
A quick parallelism check -- are any of these values that are being returned pointers? Accessing Items() is thread-safe but I'm wondering if we are holding references and if we could be affected by any other threads doing work on the same cache

Reply:
Let me double check the underlying cache implementation, but I believe the answer should be no.

Reply:
Looks like the answer is no, each time you set a value in the cache a new `Item` is created. We also create a new parameter each time we make a `Get` call to the ssm provider.